File: tests/test_cross_module_integration.py
# -*- coding: utf-8 -*-
"""
test_cross_module_integration.py — End-to-End Pipeline Monkey Tests
====================================================================

Tests the full pipeline flow: chunker → dedup → cache → reranker → detect.
Uses mocked embedding models and LLM backends.

Run:
    pytest tests/test_cross_module_integration.py -v --tb=short -x
"""

import random
import string
import sys
from pathlib import Path
from typing import Dict, List
from unittest.mock import MagicMock, patch

import pytest

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

_CHAOS_STRINGS: list[str] = [
    "",
    "   ",
    "\x00\x01\x02",
    "A" * 50_000,
    "🔥" * 2_000,
    "<script>alert('xss')</script>",
    "'; DROP TABLE users; --",
    "Hello 你好 مرحبا こんにちは",
    "\u202e\u200b\u200c\u200d\ufeff",
    "NaN",
    "null",
    "what is quantum computing?",
    "explain the theory of relativity in simple terms",
]

_SAMPLE_PARAGRAPHS = [
    "Quantum computing harnesses quantum mechanical phenomena such as superposition and entanglement to process information. Unlike classical bits that are either 0 or 1, quantum bits (qubits) can exist in multiple states simultaneously, enabling quantum computers to solve certain problems exponentially faster than classical computers.",
    "The theory of general relativity, published by Albert Einstein in 1915, describes gravity as the curvature of spacetime caused by mass and energy. This revolutionary theory replaced Newton's law of universal gravitation and predicted phenomena such as gravitational waves and black holes.",
    "Machine learning is a subset of artificial intelligence focused on building systems that learn from data. Through algorithms that iteratively learn from data, machine learning allows computers to find hidden insights without being explicitly programmed where to look.",
    "The human genome contains approximately 3 billion base pairs of DNA, organized into 23 pairs of chromosomes. The Human Genome Project, completed in 2003, was an international scientific research project with the goal of determining the sequence of all nucleotide base pairs.",
    "Photosynthesis is the process by which green plants and some other organisms use sunlight to synthesize nutrients from carbon dioxide and water. This process generates oxygen as a byproduct and is fundamental to most life on Earth.",
]


def _mock_model():
    model = MagicMock()
    model.encode.return_value = [random.random() for _ in range(384)]
    return model


# ═════════════════════════════════════════════════════════════════════════════
#  Import Chain Tests
# ═════════════════════════════════════════════════════════════════════════════


@pytest.mark.monkey
class TestImportChainMonkey:
    """Verify all pipeline modules can be imported together."""

    def test_import_chunker(self):
        from zena_mode.chunker import Chunk, ChunkerConfig, HierarchicalChunker, TextChunker

        assert TextChunker is not None
        assert HierarchicalChunker is not None

    def test_import_deduplicator(self):
        from Core.smart_deduplicator import SmartDeduplicator

        assert SmartDeduplicator is not None

    def test_import_cache(self):
        from Core.zero_waste_cache import ZeroWasteCache

        assert ZeroWasteCache is not None

    def test_import_reranker(self):
        from Core.reranker_advanced import AdvancedReranker

        assert AdvancedReranker is not None

    def test_import_hallucination_detector(self):
        from Core.hallucination_detector_v2 import AdvancedHallucinationDetector

        assert AdvancedHallucinationDetector is not None

    def test_import_confidence_scorer(self):
        from Core.confidence_scorer import AnswerQualityAssessor

        assert AnswerQualityAssessor is not None

    def test_import_conflict_detector(self):
        from Core.conflict_detector import ConflictDetector

        assert ConflictDetector is not None

    def test_import_all_at_once(self):
        """All modules loaded together — no circular import issues."""
        from zena_mode.chunker import TextChunker
        from Core.smart_deduplicator import SmartDeduplicator
        from Core.zero_waste_cache import ZeroWasteCache
        from Core.reranker_advanced import AdvancedReranker
        from Core.hallucination_detector_v2 import AdvancedHallucinationDetector
        from Core.confidence_scorer import AnswerQualityAssessor

        assert all([TextChunker, SmartDeduplicator, ZeroWasteCache, AdvancedReranker, AdvancedHallucinationDetector, AnswerQualityAssessor])


# ═════════════════════════════════════════════════════════════════════════════
#  Chunker → Deduplicator Pipeline
# ═════════════════════════════════════════════════════════════════════════════


@pytest.mark.monkey
class TestChunkerToDeduplicatorMonkey:
    """Full flow: text → chunks → dedup."""

    def test_chunk_then_dedup(self):
        from zena_mode.chunker import TextChunker
        from Core.smart_deduplicator import SmartDeduplicator

        chunker = TextChunker()
        text = " ".join(_SAMPLE_PARAGRAPHS)
        chunks = chunker.chunk_document(text)

        try:
            dedup = SmartDeduplicator()
            for chunk in chunks:
                result = dedup.check(chunk.text if hasattr(chunk, "text") else str(chunk))
                assert hasattr(result, "should_skip")
        except (TypeError, AttributeError) as e:
            # SmartDeduplicator may need different init or check signature
            pass

    def test_duplicate_text_detected(self):
        """Same text chunked twice — dedup should catch duplicates."""
        from zena_mode.chunker import TextChunker
        from Core.smart_deduplicator import SmartDeduplicator

        chunker = TextChunker()
        text = _SAMPLE_PARAGRAPHS[0] * 3
        chunks = chunker.chunk_document(text)

        try:
            dedup = SmartDeduplicator()
            skip_count = 0
            for chunk in chunks:
                result = dedup.check(chunk.text if hasattr(chunk, "text") else str(chunk))
                if result.should_skip:
                    skip_count += 1
            # At least some duplicates should be detected
        except (TypeError, AttributeError):
            pass

    def test_chaos_through_pipeline(self):
        """Chaos strings → chunker → dedup — no crashes."""
        from zena_mode.chunker import TextChunker

        chunker = TextChunker()
        for s in _CHAOS_STRINGS:
            chunks = chunker.chunk_document(s)
            assert isinstance(chunks, list)


# ═════════════════════════════════════════════════════════════════════════════
#  Full Pipeline: Chunk → Cache → Rerank
# ═════════════════════════════════════════════════════════════════════════════


@pytest.mark.monkey
class TestFullPipelineMonkey:
    """Mock-based end-to-end flow testing."""

    def test_chunk_cache_rerank_flow(self):
        """chunks → cache set → cache get → rerank."""
        from zena_mode.chunker import TextChunker
        from Core.zero_waste_cache import ZeroWasteCache

        model = _mock_model()
        chunker = TextChunker()
        cache = ZeroWasteCache(model=model, max_entries=100)

        # Chunk a document
        text = "\n\n".join(_SAMPLE_PARAGRAPHS)
        chunks = chunker.chunk_document(text)
        chunk_dicts = [{"text": c.text, "score": 0.9} for c in chunks] if chunks else []

        if chunk_dicts:
            # Cache the results
            cache.set_answer("what is quantum computing", results=chunk_dicts, source_chunks=chunk_dicts)

            # Retrieve
            result = cache.get_answer("what is quantum computing")
            # result may be None (threshold) or list — both OK

    def test_empty_pipeline_graceful(self):
        """Empty input through entire pipeline — graceful degradation."""
        from zena_mode.chunker import TextChunker
        from Core.zero_waste_cache import ZeroWasteCache

        model = _mock_model()
        chunker = TextChunker()
        cache = ZeroWasteCache(model=model, max_entries=100)

        chunks = chunker.chunk_document("")
        assert chunks == [] or isinstance(chunks, list)

        # Cache empty results
        cache.set_answer("empty query", results=[], source_chunks=[])
        result = cache.get_answer("empty query")
        assert result is None or isinstance(result, list)

    def test_adversarial_queries(self):
        """Chaos strings as queries through cached pipeline."""
        from Core.zero_waste_cache import ZeroWasteCache

        model = _mock_model()
        cache = ZeroWasteCache(model=model, max_entries=100)

        # Pre-fill with normal data
        cache.set_answer(
            "normal query",
            results=[{"text": "answer"}],
            source_chunks=[{"text": "chunk"}],
        )

        # Query with adversarial strings
        for s in _CHAOS_STRINGS:
            try:
                result = cache.get_answer(s)
                assert result is None or isinstance(result, (list, tuple, dict))
            except (TypeError, ValueError):
                pass  # rejecting garbage is fine


# ═════════════════════════════════════════════════════════════════════════════
#  Graceful Degradation Tests
# ═════════════════════════════════════════════════════════════════════════════


@pytest.mark.monkey
class TestGracefulDegradationMonkey:
    """Pipeline with disabled/missing optional modules."""

    def test_chunker_alone(self):
        """Chunker works standalone without any other module."""
        from zena_mode.chunker import TextChunker

        chunker = TextChunker()
        result = chunker.chunk_document("Hello world, this is a standalone test.")
        assert isinstance(result, list)

    def test_cache_alone(self):
        """Cache works standalone without chunker or reranker."""
        from Core.zero_waste_cache import ZeroWasteCache

        cache = ZeroWasteCache(model=_mock_model(), max_entries=10)
        cache.set_answer("q", results=[{"text": "a"}], source_chunks=[{"text": "c"}])
        result = cache.get_answer("q")
        # Any result is fine

    def test_dedup_alone(self):
        """SmartDeduplicator works standalone."""
        from Core.smart_deduplicator import SmartDeduplicator

        try:
            dedup = SmartDeduplicator()
            result = dedup.check("test text for deduplication")
            assert hasattr(result, "should_skip")
        except (TypeError, AttributeError):
            pass  # may need model param

    def test_modules_dont_depend_on_qdrant(self):
        """Core modules should not crash if qdrant_client is not available."""
        # These modules should be importable regardless of Qdrant
        from Core.zero_waste_cache import ZeroWasteCache
        from Core.smart_deduplicator import SmartDeduplicator
        from Core.confidence_scorer import AnswerQualityAssessor
        from Core.hallucination_detector_v2 import AdvancedHallucinationDetector
        from zena_mode.chunker import TextChunker

        assert all([ZeroWasteCache, SmartDeduplicator, AnswerQualityAssessor, AdvancedHallucinationDetector, TextChunker])

File: rust_output/src/test_cross_module_integration.rs
/// test_cross_module_integration::py — End-to-End Pipeline Monkey Tests
/// ====================================================================
/// 
/// Tests the full pipeline flow: chunker → dedup → cache → reranker → detect.
/// Uses mocked embedding models and LLM backends.
/// 
/// Run:
/// pytest tests/test_cross_module_integration::py -v --tb=short -x

use anyhow::{Result, Context};
use std::collections::HashMap;

pub const ROOT: &str = "Path(file!()).resolve().parent.parent";

pub static _CHAOS_STRINGS: std::sync::LazyLock<Vec<String>> = std::sync::LazyLock::new(|| Vec::new());

pub static _SAMPLE_PARAGRAPHS: std::sync::LazyLock<Vec<String>> = std::sync::LazyLock::new(|| Vec::new());

/// Verify all pipeline modules can be imported together.
#[derive(Debug, Clone)]
pub struct TestImportChainMonkey {
}

impl TestImportChainMonkey {
    pub fn test_import_chunker(&self) -> () {
        // TODO: from zena_mode.chunker import Chunk, ChunkerConfig, HierarchicalChunker, TextChunker
        assert!(TextChunker.is_some());
        assert!(HierarchicalChunker.is_some());
    }
    pub fn test_import_deduplicator(&self) -> () {
        // TODO: from Core.smart_deduplicator import SmartDeduplicator
        assert!(SmartDeduplicator.is_some());
    }
    pub fn test_import_cache(&self) -> () {
        // TODO: from Core.zero_waste_cache import ZeroWasteCache
        assert!(ZeroWasteCache.is_some());
    }
    pub fn test_import_reranker(&self) -> () {
        // TODO: from Core.reranker_advanced import AdvancedReranker
        assert!(AdvancedReranker.is_some());
    }
    pub fn test_import_hallucination_detector(&self) -> () {
        // TODO: from Core.hallucination_detector_v2 import AdvancedHallucinationDetector
        assert!(AdvancedHallucinationDetector.is_some());
    }
    pub fn test_import_confidence_scorer(&self) -> () {
        // TODO: from Core.confidence_scorer import AnswerQualityAssessor
        assert!(AnswerQualityAssessor.is_some());
    }
    pub fn test_import_conflict_detector(&self) -> () {
        // TODO: from Core.conflict_detector import ConflictDetector
        assert!(ConflictDetector.is_some());
    }
    /// All modules loaded together — no circular import issues.
    pub fn test_import_all_at_once(&self) -> () {
        // All modules loaded together — no circular import issues.
        // TODO: from zena_mode.chunker import TextChunker
        // TODO: from Core.smart_deduplicator import SmartDeduplicator
        // TODO: from Core.zero_waste_cache import ZeroWasteCache
        // TODO: from Core.reranker_advanced import AdvancedReranker
        // TODO: from Core.hallucination_detector_v2 import AdvancedHallucinationDetector
        // TODO: from Core.confidence_scorer import AnswerQualityAssessor
        assert!(vec![TextChunker, SmartDeduplicator, ZeroWasteCache, AdvancedReranker, AdvancedHallucinationDetector, AnswerQualityAssessor].iter().all(|v| *v));
    }
}

/// Full flow: text → chunks → dedup.
#[derive(Debug, Clone)]
pub struct TestChunkerToDeduplicatorMonkey {
}

impl TestChunkerToDeduplicatorMonkey {
    pub fn test_chunk_then_dedup(&self) -> Result<()> {
        // TODO: from zena_mode.chunker import TextChunker
        // TODO: from Core.smart_deduplicator import SmartDeduplicator
        let mut chunker = TextChunker();
        let mut text = _SAMPLE_PARAGRAPHS.join(&" ".to_string());
        let mut chunks = chunker::chunk_document(text);
        // try:
        {
            let mut dedup = SmartDeduplicator();
            for chunk in chunks.iter() {
                let mut result = dedup::check(if /* hasattr(chunk, "text".to_string()) */ true { chunk.text } else { chunk.to_string() });
                assert!(/* hasattr(result, "should_skip".to_string()) */ true);
            }
        }
        // except (TypeError, AttributeError) as e:
    }
    /// Same text chunked twice — dedup should catch duplicates.
    pub fn test_duplicate_text_detected(&self) -> Result<()> {
        // Same text chunked twice — dedup should catch duplicates.
        // TODO: from zena_mode.chunker import TextChunker
        // TODO: from Core.smart_deduplicator import SmartDeduplicator
        let mut chunker = TextChunker();
        let mut text = (_SAMPLE_PARAGRAPHS[0] * 3);
        let mut chunks = chunker::chunk_document(text);
        // try:
        {
            let mut dedup = SmartDeduplicator();
            let mut skip_count = 0;
            for chunk in chunks.iter() {
                let mut result = dedup::check(if /* hasattr(chunk, "text".to_string()) */ true { chunk.text } else { chunk.to_string() });
                if result.should_skip {
                    skip_count += 1;
                }
            }
        }
        // except (TypeError, AttributeError) as _e:
    }
    /// Chaos strings → chunker → dedup — no crashes.
    pub fn test_chaos_through_pipeline(&self) -> () {
        // Chaos strings → chunker → dedup — no crashes.
        // TODO: from zena_mode.chunker import TextChunker
        let mut chunker = TextChunker();
        for s in _CHAOS_STRINGS.iter() {
            let mut chunks = chunker::chunk_document(s);
            assert!(/* /* isinstance(chunks, list) */ */ true);
        }
    }
}

/// Mock-based end-to-end flow testing.
#[derive(Debug, Clone)]
pub struct TestFullPipelineMonkey {
}

impl TestFullPipelineMonkey {
    /// chunks → cache set → cache get → rerank.
    pub fn test_chunk_cache_rerank_flow(&self) -> () {
        // chunks → cache set → cache get → rerank.
        // TODO: from zena_mode.chunker import TextChunker
        // TODO: from Core.zero_waste_cache import ZeroWasteCache
        let mut model = _mock_model();
        let mut chunker = TextChunker();
        let mut cache = ZeroWasteCache(/* model= */ model, /* max_entries= */ 100);
        let mut text = _SAMPLE_PARAGRAPHS.join(&"\n\n".to_string());
        let mut chunks = chunker::chunk_document(text);
        let mut chunk_dicts = if chunks { chunks.iter().map(|c| HashMap::from([("text".to_string(), c.text), ("score".to_string(), 0.9_f64)])).collect::<Vec<_>>() } else { vec![] };
        if chunk_dicts {
            cache::set_answer("what is quantum computing".to_string(), /* results= */ chunk_dicts, /* source_chunks= */ chunk_dicts);
            let mut result = cache::get_answer("what is quantum computing".to_string());
        }
    }
    /// Empty input through entire pipeline — graceful degradation.
    pub fn test_empty_pipeline_graceful(&self) -> () {
        // Empty input through entire pipeline — graceful degradation.
        // TODO: from zena_mode.chunker import TextChunker
        // TODO: from Core.zero_waste_cache import ZeroWasteCache
        let mut model = _mock_model();
        let mut chunker = TextChunker();
        let mut cache = ZeroWasteCache(/* model= */ model, /* max_entries= */ 100);
        let mut chunks = chunker::chunk_document("".to_string());
        assert!((chunks == vec![] || /* /* isinstance(chunks, list) */ */ true));
        cache::set_answer("empty query".to_string(), /* results= */ vec![], /* source_chunks= */ vec![]);
        let mut result = cache::get_answer("empty query".to_string());
        assert!((result.is_none() || /* /* isinstance(result, list) */ */ true));
    }
    /// Chaos strings as queries through cached pipeline.
    pub fn test_adversarial_queries(&self) -> Result<()> {
        // Chaos strings as queries through cached pipeline.
        // TODO: from Core.zero_waste_cache import ZeroWasteCache
        let mut model = _mock_model();
        let mut cache = ZeroWasteCache(/* model= */ model, /* max_entries= */ 100);
        cache::set_answer("normal query".to_string(), /* results= */ vec![HashMap::from([("text".to_string(), "answer".to_string())])], /* source_chunks= */ vec![HashMap::from([("text".to_string(), "chunk".to_string())])]);
        for s in _CHAOS_STRINGS.iter() {
            // try:
            {
                let mut result = cache::get_answer(s);
                assert!((result.is_none() || /* /* isinstance(result, (list, tuple, dict) */) */ true));
            }
            // except (TypeError, ValueError) as _e:
        }
    }
}

/// Pipeline with disabled/missing optional modules.
#[derive(Debug, Clone)]
pub struct TestGracefulDegradationMonkey {
}

impl TestGracefulDegradationMonkey {
    /// Chunker works standalone without any other module.
    pub fn test_chunker_alone(&self) -> () {
        // Chunker works standalone without any other module.
        // TODO: from zena_mode.chunker import TextChunker
        let mut chunker = TextChunker();
        let mut result = chunker::chunk_document("Hello world, this is a standalone test.".to_string());
        assert!(/* /* isinstance(result, list) */ */ true);
    }
    /// Cache works standalone without chunker or reranker.
    pub fn test_cache_alone(&self) -> () {
        // Cache works standalone without chunker or reranker.
        // TODO: from Core.zero_waste_cache import ZeroWasteCache
        let mut cache = ZeroWasteCache(/* model= */ _mock_model(), /* max_entries= */ 10);
        cache::set_answer("q".to_string(), /* results= */ vec![HashMap::from([("text".to_string(), "a".to_string())])], /* source_chunks= */ vec![HashMap::from([("text".to_string(), "c".to_string())])]);
        let mut result = cache::get_answer("q".to_string());
    }
    /// SmartDeduplicator works standalone.
    pub fn test_dedup_alone(&self) -> Result<()> {
        // SmartDeduplicator works standalone.
        // TODO: from Core.smart_deduplicator import SmartDeduplicator
        // try:
        {
            let mut dedup = SmartDeduplicator();
            let mut result = dedup::check("test text for deduplication".to_string());
            assert!(/* hasattr(result, "should_skip".to_string()) */ true);
        }
        // except (TypeError, AttributeError) as _e:
    }
    /// Core modules should not crash if qdrant_client is not available.
    pub fn test_modules_dont_depend_on_qdrant(&self) -> () {
        // Core modules should not crash if qdrant_client is not available.
        // TODO: from Core.zero_waste_cache import ZeroWasteCache
        // TODO: from Core.smart_deduplicator import SmartDeduplicator
        // TODO: from Core.confidence_scorer import AnswerQualityAssessor
        // TODO: from Core.hallucination_detector_v2 import AdvancedHallucinationDetector
        // TODO: from zena_mode.chunker import TextChunker
        assert!(vec![ZeroWasteCache, SmartDeduplicator, AnswerQualityAssessor, AdvancedHallucinationDetector, TextChunker].iter().all(|v| *v));
    }
}

pub fn _mock_model() -> () {
    let mut model = MagicMock();
    model.encode.return_value = 0..384.iter().map(|_| random.random()).collect::<Vec<_>>();
    model
}
